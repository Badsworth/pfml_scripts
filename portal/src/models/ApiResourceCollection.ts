export interface ApiResource {
  // API resources are expected to be objects. We expect at least an ID property.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any;
}

/**
 * Readonly generic class representing a collection of a specific data model. Methods
 * return new instances of the collection rather than modifying the original collection.
 * @example const c = new ApiResourceCollection<Document>("document_id", [{ document_id: "123", name: "example" }]);
 *          c.setItem({ document_id: "456", name: "example2" });
 */
class ApiResourceCollection<TApiResource extends ApiResource> {
  idKey: keyof TApiResource;
  private map: Map<string, TApiResource>;

  constructor(idKey: keyof TApiResource, items: TApiResource[] = []) {
    this.idKey = idKey;
    this.map = new Map(
      items.map((item) => {
        return [item[idKey], item];
      })
    );
  }

  private get _mutableMap() {
    return new Map(this.map);
  }

  private _arrayFromMap(map: Map<string, TApiResource>): TApiResource[] {
    return Array.from(map.values());
  }

  get items(): TApiResource[] {
    return this._arrayFromMap(this.map);
  }

  get isEmpty(): boolean {
    return this.map.size === 0;
  }

  getItem(itemId: string): TApiResource | undefined {
    return this.map.get(itemId);
  }

  /**
   * Add or update an item in the collection. Returns a new collection with the
   * item included. Does not modify the original collection.
   */
  setItem(item: TApiResource): ApiResourceCollection<TApiResource> {
    const updatedMap = this._mutableMap.set(item[this.idKey], item);
    return new ApiResourceCollection<TApiResource>(
      this.idKey,
      this._arrayFromMap(updatedMap)
    );
  }

  /**
   * Add or update items in the collection. Returns a new collection with the
   * items included. Does not modify the original collection.
   */
  setItems(items: TApiResource[] = []): ApiResourceCollection<TApiResource> {
    const updatedMap = this._mutableMap;
    items.forEach((item) => {
      updatedMap.set(item[this.idKey], item);
    });
    return new ApiResourceCollection<TApiResource>(
      this.idKey,
      this._arrayFromMap(updatedMap)
    );
  }

  /**
   * Removes an item from the collection. Returns a new collection with the item removed.
   * Does not modify the original collection.
   */
  removeItem(itemId: string): ApiResourceCollection<TApiResource> {
    const updatedMap = this._mutableMap;
    updatedMap.delete(itemId);

    return new ApiResourceCollection<TApiResource>(
      this.idKey,
      this._arrayFromMap(updatedMap)
    );
  }
}

export default ApiResourceCollection;
