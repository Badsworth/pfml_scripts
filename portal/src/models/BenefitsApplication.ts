/**
 * @file Benefits application model and enum values
 */

import LeaveReason, { LeaveReasonType } from "./LeaveReason";
import { compact, get, merge, sum, sumBy, zip } from "lodash";

import Address from "./Address";
import BaseBenefitsApplication from "./BaseBenefitsApplication";
import ConcurrentLeave from "./ConcurrentLeave";
import EmployerBenefit from "./EmployerBenefit";
import OrganizationUnit from "./OrganizationUnit";
import OtherIncome from "./OtherIncome";
import PaymentPreference from "./PaymentPreference";
import PreviousLeave from "./PreviousLeave";
import { ValuesOf } from "../../types/common";
import assert from "assert";
import dayjs from "dayjs";
import isBlank from "../utils/isBlank";
import spreadMinutesOverWeek from "../utils/spreadMinutesOverWeek";

class BenefitsApplication extends BaseBenefitsApplication {
  application_id: string;
  fineos_absence_id: string | null = null;
  organization_unit_id: string | null = null;
  organization_unit_selection: "not_listed" | "not_selected" | null = null;

  first_name: string | null = null;
  middle_name: string | null = null;
  last_name: string | null = null;
  computed_earliest_submission_date: string | null = null;
  concurrent_leave: ConcurrentLeave | null = null;
  employer_benefits: EmployerBenefit[] = [];
  date_of_birth: string | null = null;
  employee_id: string | null = null;
  employer_fein: string | null = null;
  gender: ValuesOf<typeof Gender> | null = null;
  has_concurrent_leave: boolean | null = null;
  has_continuous_leave_periods: boolean | null = null;
  has_employer_benefits: boolean | null = null;
  has_intermittent_leave_periods: boolean | null = null;
  has_mailing_address: boolean | null = null;
  has_other_incomes: boolean | null = null;
  has_previous_leaves_other_reason: boolean | null = null;
  has_previous_leaves_same_reason: boolean | null = null;
  has_reduced_schedule_leave_periods: boolean | null = null;
  has_state_id: boolean | null = null;
  has_submitted_payment_preference: boolean | null = null;
  hours_worked_per_week: number | null = null;
  is_withholding_tax: boolean | null = null;
  mass_id: string | null = null;
  mailing_address: Address | null = null;
  other_incomes: OtherIncome[] = [];
  payment_preference: PaymentPreference | null = null;
  previous_leaves_other_reason: PreviousLeave[] = [];
  previous_leaves_same_reason: PreviousLeave[] = [];
  residential_address: Address = new Address({});
  tax_identifier: string | null = null;
  work_pattern: Partial<WorkPattern> | null = null;

  employment_status: ValuesOf<typeof EmploymentStatus> | null = null;

  computed_start_dates: {
    other_reason?: string | null;
    same_reason?: string | null;
  };

  leave_details: {
    continuous_leave_periods: ContinuousLeavePeriod[];
    intermittent_leave_periods: IntermittentLeavePeriod[];
    reduced_schedule_leave_periods: ReducedScheduleLeavePeriod[];
    employer_notified: boolean | null;
    employer_notification_date: string | null;
    caring_leave_metadata: CaringLeaveMetadata | null;
    child_birth_date: string | null;
    child_placement_date: string | null;
    has_future_child_date: boolean | null;
    pregnant_or_recent_birth: boolean | null;
    reason: ValuesOf<typeof LeaveReason> | null;
    reason_qualifier: ReasonQualifierEnum | null;
  };

  phone: {
    int_code: string | null;
    phone_number: string | null;
    phone_type: ValuesOf<typeof PhoneType> | null;
  };

  status: ValuesOf<typeof BenefitsApplicationStatus> | null = null;

  // Organization unit data selected by the user (used in review page)
  organization_unit: OrganizationUnit | null;

  // The list of organization units that we know are connected
  // to this employee based on their occupation and DUA data
  employee_organization_units: OrganizationUnit[] = [];

  // The list of organization units that are connected to this employer (in FINEOS)
  employer_organization_units: OrganizationUnit[] = [];

  constructor(attrs: Partial<BenefitsApplication>) {
    super();
    // Recursively merge with the defaults
    merge(this, attrs);
  }

  /**
   * Applications imported from Fineos as part of Channel Switching won't have
   * any caring leave metadata fields set.
   */
  get hasCaringLeaveMetadata(): boolean {
    return !isBlank(
      this.leave_details.caring_leave_metadata?.family_member_first_name
    );
  }

  /**
   * Determine if applicable leave period start date(s) are in the future.
   */
  get isLeaveStartDateInFuture() {
    const startDates: string[] = compact([
      get(this, "leave_details.continuous_leave_periods[0].start_date"),
      get(this, "leave_details.intermittent_leave_periods[0].start_date"),
      get(this, "leave_details.reduced_schedule_leave_periods[0].start_date"),
    ]);

    if (!startDates.length) return false;

    const now = dayjs().format("YYYY-MM-DD"); // current date in ISO 8601 string in local time (by default day.js parses in local time)
    return startDates.every((startDate) => {
      // Compare the two dates lexicographically. This works since they're both in
      // ISO-8601 format, eg "2020-10-13"
      return startDate > now;
    });
  }

  /**
   * Determine if the earliest possible application submission date is in the future (i.e., they must wait to submit their application).
   */
  get isEarliestSubmissionDateInFuture() {
    if (this.computed_earliest_submission_date === null) {
      return false;
    }

    const now = dayjs().format("YYYY-MM-DD");
    return this.computed_earliest_submission_date > now;
  }

  /**
   * Check if Claim has been marked as completed yet.
   */
  get isCompleted() {
    return this.status === BenefitsApplicationStatus.completed;
  }

  /**
   * Determine if claim is a Medical or Pregnancy leave claim
   */
  get isMedicalOrPregnancyLeave(): boolean {
    const reason: LeaveReasonType = get(this, "leave_details.reason");
    return reason === LeaveReason.medical || reason === LeaveReason.pregnancy;
  }

  /**
   * Determine if claim is a Caring Leave claim
   */
  get isCaringLeave(): boolean {
    return get(this, "leave_details.reason") === LeaveReason.care;
  }

  /**
   * Check if Claim has been submitted yet. This affects the editability
   * of some fields, and as a result, the user experience.
   */
  get isSubmitted() {
    return (
      this.status === BenefitsApplicationStatus.submitted ||
      this.status === BenefitsApplicationStatus.completed
    );
  }

  /**
   * Returns a list of the employer's organization units
   * except for the ones connected to the employee
   */
  get extraOrgUnits() {
    return this.employer_organization_units.filter(
      (o) =>
        !this.employee_organization_units
          .map((o) => o.organization_unit_id)
          .includes(o.organization_unit_id)
    );
  }

  /**
   * Returns earliest start date across all leave periods
   */
  get leaveStartDate() {
    const periods = [
      get(this, "leave_details.continuous_leave_periods"),
      get(this, "leave_details.intermittent_leave_periods"),
      get(this, "leave_details.reduced_schedule_leave_periods"),
    ].flat();

    const startDates: string[] = compact(periods)
      .map((period) => period.start_date)
      .sort();

    if (!startDates.length) return null;

    return startDates[0];
  }

  /**
   * Returns latest end date across all leave periods
   */
  get leaveEndDate() {
    const periods = [
      get(this, "leave_details.continuous_leave_periods"),
      get(this, "leave_details.intermittent_leave_periods"),
      get(this, "leave_details.reduced_schedule_leave_periods"),
    ].flat();

    const endDates: string[] = compact(periods)
      .map((period) => period.end_date)
      .sort();

    if (!endDates.length) return null;

    return endDates[endDates.length - 1];
  }
}

/**
 * Enums for the Application's `status` field
 * @enum {string}
 */
export const BenefitsApplicationStatus = {
  // Submitted by the user (Step 2)
  completed: "Completed",
  // AKA: In Progress (Step 1)
  started: "Started",
  // Stored in the claims processing system (Step 3)
  submitted: "Submitted",
  // User not found - No link between employee and employer
  // User completed Step 1 and part of Step 2 along with additional information to help make a manual match
  inReview: "In Review",
} as const;

/**
 * Enums for the Application's `employment_status` field
 * @enum {string}
 */
export const EmploymentStatus = {
  employed: "Employed",
  selfEmployed: "Self-Employed",
  unemployed: "Unemployed",
} as const;

/**
 * Enums for the Application's `leave_details.reason_qualifier` field
 * @enum {string}
 */
export const ReasonQualifier = {
  adoption: "Adoption",
  fosterCare: "Foster Care",
  newBorn: "Newborn",
} as const;

export type ReasonQualifierEnum = ValuesOf<typeof ReasonQualifier>;

export class ContinuousLeavePeriod {
  leave_period_id: string | null = null;
  end_date: string | null = null;
  start_date: string | null = null;

  constructor(attrs: Partial<ContinuousLeavePeriod>) {
    Object.assign(this, attrs);
  }
}

export class IntermittentLeavePeriod {
  leave_period_id: string | null = null;
  start_date: string | null = null;
  end_date: string | null = null;
  // How many {days|hours} of work will you miss per absence?
  duration: number | null = null;
  // How long will an absence typically last?
  duration_basis: ValuesOf<typeof DurationBasis> | null = null;

  // Estimate how many absences {per week|per month|over the next 6 months}
  frequency: number | null = null;
  // Implied by input selection of "over the next 6 months"
  // and can only ever be equal to 6
  frequency_interval: number | null = null;
  // How often might you need to be absent from work?
  frequency_interval_basis: ValuesOf<typeof FrequencyIntervalBasis> | null =
    null;

  constructor(attrs: Partial<IntermittentLeavePeriod>) {
    Object.assign(this, attrs);
  }
}

export class CaringLeaveMetadata {
  family_member_date_of_birth: string | null = null;
  family_member_first_name: string | null = null;
  family_member_last_name: string | null = null;
  family_member_middle_name: string | null = null;
  relationship_to_caregiver: ValuesOf<typeof RelationshipToCaregiver> | null =
    null;

  constructor(attrs: Partial<CaringLeaveMetadata>) {
    Object.assign(this, attrs);
  }
}

export class WorkPattern {
  work_pattern_days: WorkPatternDay[] | null = [];
  work_pattern_type: ValuesOf<typeof WorkPatternType> | null = null;

  constructor(attrs: Partial<WorkPattern>) {
    Object.assign(this, attrs);

    if (!this.work_pattern_days || !this.work_pattern_days.length) {
      this.work_pattern_days = OrderedDaysOfWeek.map(
        (day_of_week) => new WorkPatternDay({ day_of_week, minutes: null })
      );
    }

    assert(
      this.work_pattern_days.length === 7,
      `${this.work_pattern_days.length} work_pattern_days length must be 7. Consider using WorkPattern's static createWithWeek.`
    );
  }

  /**
   * Return total minutes worked for work pattern days. Returns null if no minutes are defined for work pattern days
   */
  get minutesWorkedPerWeek() {
    const hasNoMinutes =
      this.work_pattern_days &&
      this.work_pattern_days.every((day) => day.minutes === null);
    if (hasNoMinutes) {
      return null;
    }

    return sumBy(this.work_pattern_days, "minutes");
  }

  /**
   * Create a WorkPattern with a week, splitting provided minutes across 7 work_pattern_days.
   * @param minutesWorkedPerWeek - average minutes worked per week. Must be an integer. Will split minutes evenly across 7 day week
   * @param workPattern - work pattern attributes to apply to new WorkPattern
   */
  static createWithWeek(
    minutesWorkedPerWeek: number,
    workPattern: WorkPattern | { [key: string]: never } = {}
  ) {
    const minutesOverWeek = spreadMinutesOverWeek(minutesWorkedPerWeek);

    const newWeek = zip(OrderedDaysOfWeek, minutesOverWeek).map(
      ([day_of_week, minutes]) =>
        new WorkPatternDay({
          day_of_week,
          minutes,
        })
    );

    return new WorkPattern({
      ...workPattern,
      work_pattern_days: newWeek,
    });
  }
}

export class WorkPatternDay {
  day_of_week: typeof OrderedDaysOfWeek[number] | null = null;
  // API represents hours in minutes
  minutes: number | null = null;

  constructor(attrs: Partial<WorkPatternDay>) {
    Object.assign(this, attrs);
  }
}

/**
 * Enums for the Application's `intermittent_leave_periods[].frequency_interval_basis` field
 * @enum {string}
 */
export const FrequencyIntervalBasis = {
  // Days is also a valid enum in the API, however the Portal
  // doesn't offer this as an option to the user
  // days: "Days",
  months: "Months",
  weeks: "Weeks",
} as const;

/**
 * Enums for the Application's `intermittent_leave_periods[].duration_basis` field
 * @enum {string}
 */
export const DurationBasis = {
  days: "Days",
  hours: "Hours",
  // Minutes is also a valid enum in the API, however the Portal
  // doesn't offer this as an option to the user
  // minutes: "Minutes",
} as const;

export class ReducedScheduleLeavePeriod {
  leave_period_id: string | null = null;
  start_date: string | null = null;
  end_date: string | null = null;
  sunday_off_minutes: number | null = null;
  monday_off_minutes: number | null = null;
  tuesday_off_minutes: number | null = null;
  wednesday_off_minutes: number | null = null;
  thursday_off_minutes: number | null = null;
  friday_off_minutes: number | null = null;
  saturday_off_minutes: number | null = null;

  constructor(attrs: Partial<ReducedScheduleLeavePeriod>) {
    Object.assign(this, attrs);
  }

  get days() {
    return [
      {
        day_of_week: DayOfWeek.sunday,
        minutes: this.sunday_off_minutes,
      },
      {
        day_of_week: DayOfWeek.monday,
        minutes: this.monday_off_minutes,
      },
      {
        day_of_week: DayOfWeek.tuesday,
        minutes: this.tuesday_off_minutes,
      },
      {
        day_of_week: DayOfWeek.wednesday,
        minutes: this.wednesday_off_minutes,
      },
      {
        day_of_week: DayOfWeek.thursday,
        minutes: this.thursday_off_minutes,
      },
      {
        day_of_week: DayOfWeek.friday,
        minutes: this.friday_off_minutes,
      },
      {
        day_of_week: DayOfWeek.saturday,
        minutes: this.saturday_off_minutes,
      },
    ];
  }

  /**
   */
  get totalMinutesOff() {
    const fieldsWithMinutes = compact([
      this.friday_off_minutes,
      this.monday_off_minutes,
      this.saturday_off_minutes,
      this.sunday_off_minutes,
      this.thursday_off_minutes,
      this.tuesday_off_minutes,
      this.wednesday_off_minutes,
    ]);

    if (fieldsWithMinutes.length) {
      return sum(fieldsWithMinutes);
    }

    return null;
  }
}

/**
 * Enums for the Application's `phone.phone_type` field
 * @enum {string}
 */
export const PhoneType = {
  cell: "Cell",
  phone: "Phone",
} as const;

/* Enums for claimant gender options
 * @enum {string}
 */
export const Gender = {
  genderNotListed: "Gender not listed",
  man: "Man",
  nonbinary: "Non-binary",
  preferNotToAnswer: "Prefer not to answer",
  woman: "Woman",
} as const;

/* Enums for caring leave's relationship to caregiver
 * @enum {string}
 */
export const RelationshipToCaregiver = {
  child: "Child",
  grandchild: "Grandchild",
  grandparent: "Grandparent",
  inlaw: "Inlaw",
  other: "Other",
  otherFamilyMember: "Other Family Member",
  parent: "Parent",
  serviceMember: "Service Member",
  sibling: "Sibling - Brother/Sister",
  spouse: "Spouse",
} as const;

/**
 * Enums for the Application's `work_pattern.work_pattern_type` field
 * @enum {string}
 */
export const WorkPatternType = {
  fixed: "Fixed",
  rotating: "Rotating",
  variable: "Variable",
} as const;

/**
 * Ordered days of the week
 */
export const OrderedDaysOfWeek = [
  "Sunday",
  "Monday",
  "Tuesday",
  "Wednesday",
  "Thursday",
  "Friday",
  "Saturday",
] as const;

/**
 * Enums for the Application's `work_pattern.work_pattern_days[].day_of_week` fields
 * Produces object: { sunday: "Sunday", monday: "Monday", ... }
 * @example DayOfWeek.sunday
 * @enum {string}
 */
/* eslint-disable sort-keys */
export const DayOfWeek = {
  sunday: "Sunday",
  monday: "Monday",
  tuesday: "Tuesday",
  wednesday: "Wednesday",
  thursday: "Thursday",
  friday: "Friday",
  saturday: "Saturday",
} as const;

export default BenefitsApplication;
