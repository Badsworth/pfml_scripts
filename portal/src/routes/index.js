/* eslint sort-keys: ["error", "asc"] */
/**
 * @file Routes to various pages in the application are defined here rather than being
 * hard-coded into various files.
 */
const routes = {
  api: {
    claims: "/applications",
    users: "/users",
  },
  applications: "/applications",
  auth: {
    createAccount: "/create-account",
    forgotPassword: "/forgot-password",
    login: "/login",
    resetPassword: "/reset-password",
    verifyAccount: "/verify-account",
  },
  claims: {
    averageWorkHours: "/claims/average-work-hours",
    checklist: "/claims/checklist",
    confirm: "/claims/confirm",
    dashboard: "/",
    dateOfBirth: "/claims/date-of-birth",
    duration: "/claims/duration",
    employerBenefitDetails: "/claims/employer-benefit-details",
    employerBenefits: "/claims/employer-benefits",
    employmentStatus: "/claims/employment-status",
    leaveDates: "/claims/leave-dates",
    leaveReason: "/claims/leave-reason",
    name: "/claims/name",
    notifiedEmployer: "/claims/notified-employer",
    otherIncomes: "/claims/other-incomes",
    otherIncomesDetails: "/claims/other-incomes-details",
    paymentMethod: "/claims/payment-method",
    previousLeaves: "/claims/previous-leaves",
    previousLeavesDetails: "/claims/previous-leaves-details",
    reasonPregnancy: "/claims/reason-pregnancy",
    review: "/claims/review",
    ssn: "/claims/ssn",
    stateId: "/claims/state-id",
    success: "/claims/success",
    // For routes that don't have a page to point to yet, we can route them
    // to a placeholder page. This allows us to search our code for routes.claims.todo,
    // which is less confusing than seeing routes.claims.checklist
    todo: "/claims/checklist",
    uploadHealthcareForm: "/claims/upload-healthcare-form",
    uploadOtherId: "/claims/upload-other-id",
    uploadStateId: "/claims/upload-state-id",
  },
  external: {
    massgov: {
      healthcareProviderForm:
        "https://www.mass.gov/how-to/download-a-healthcare-provider-form",
      informedConsent: "https://www.mass.gov/paidleave-informedconsent",
      privacyPolicy: "https://www.mass.gov/privacypolicy",
    },
  },
  home: "/",
  user: {
    consentToDataSharing: "/user/consent-to-data-sharing",
  },
};

export default routes;
