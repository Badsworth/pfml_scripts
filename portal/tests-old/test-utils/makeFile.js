/**
 * Make a File instance
 * @see https://developer.mozilla.org/en-US/docs/Web/API/File
 * @param {object} attrs File attributes
 * @param {string} attrs.name The file's name
 * @param {string} attrs.type The file's MIME type
 * @returns {File}
 */
export const makeFile = (attrs = {}) => {
  const { name, type } = Object.assign(
    {
      name: "file.pdf",
      type: "application/pdf",
    },
    attrs
  );

  return new File([], name, { type });
};

/**
 * Create an HTML <input> for a test's event.target
 * @param {object} attrs - HTML attributes for the input
 * @returns {HTMLInputElement}
 */
export function createInputElement(attrs) {
  const input = document.createElement("input");

  Object.entries(attrs).forEach(([attributeName, value]) => {
    input.setAttribute(attributeName, value);
  });

  if (attrs.checked === false) input.checked = false;

  return input;
}
