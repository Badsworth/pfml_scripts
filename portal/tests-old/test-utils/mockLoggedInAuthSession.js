import { Auth } from "@aws-amplify/auth";

/**
 * Mock the Cognito authentication session so a test user appears logged in
 */
const mockLoggedInAuthSession = () => {
  const accessTokenJwt =
    "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoiQnVkIn0.YDRecdsqG_plEwM0H8rK7t2z0R3XRNESJB5ZXk-FRN8";

  jest.spyOn(Auth, "currentSession").mockImplementation(() =>
    Promise.resolve({
      accessToken: { jwtToken: accessTokenJwt },
    })
  );
};

export default mockLoggedInAuthSession;
