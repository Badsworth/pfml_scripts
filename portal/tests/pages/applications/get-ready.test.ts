import GetReady, {
  GetReadyProps,
} from "../../../src/pages/applications/get-ready";
import { MockBenefitsApplicationBuilder, renderPage } from "../../test-utils";
import ApiResourceCollection from "src/models/ApiResourceCollection";
import BenefitsApplication from "../../../src/models/BenefitsApplication";
import routes from "../../../src/routes";
import { screen } from "@testing-library/react";

const setup = (
  claims: BenefitsApplication[] = [],
  queryParams?: GetReadyProps["query"]
) => {
  return renderPage(
    GetReady,
    {
      pathname: routes.applications.getReady,
      addCustomSetup: (appLogic) => {
        appLogic.benefitsApplications.benefitsApplications =
          new ApiResourceCollection<BenefitsApplication>(
            "application_id",
            claims
          );
        appLogic.benefitsApplications.loadPage = jest.fn();
      },
    },
    { query: queryParams }
  );
};

describe("GetReady", () => {
  beforeEach(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date(2022, 1, 1));
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it("renders get ready content", () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it("doesn't show link to applications when claims do not exist", () => {
    setup();
    expect(
      screen.queryByRole("link", { name: "View all applications" })
    ).not.toBeInTheDocument();
  });

  it("shows link to applications when claims exist", () => {
    const claims = [new MockBenefitsApplicationBuilder().create()];
    setup(claims);
    expect(
      screen.getByRole("link", { name: "View all applications" })
    ).toBeInTheDocument();
  });

  it("displays success alert when user sets up SMS MFA", () => {
    setup([], { smsMfaConfirmed: "true" });
    expect(screen.getAllByRole("region")[0]).toMatchSnapshot();
  });
});
