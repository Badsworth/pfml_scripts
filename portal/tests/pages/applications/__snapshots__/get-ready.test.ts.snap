// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetReady displays success alert when user sets up SMS MFA 1`] = `
<div
  aria-labelledby="alert-heading4"
  class="usa-alert usa-alert--success margin-bottom-3"
  role="region"
  tabindex="-1"
>
  <div
    class="usa-alert__body"
  >
    <h2
      class="usa-alert__heading font-heading-md text-bold"
      id="alert-heading4"
    >
      Phone number confirmed
    </h2>
    <div
      class="usa-alert__text"
    >
      To protect your security, we'll send you a 6-digit code when you log in to verify it's really you.
    </div>
  </div>
</div>
`;

exports[`GetReady renders get ready content 1`] = `
<div>
  <h1
    class="js-title margin-top-0 margin-bottom-2 font-heading-lg line-height-sans-2"
    tabindex="-1"
  >
    Get ready to apply
  </h1>
  <div
    aria-labelledby="alert-heading1"
    class="usa-alert usa-alert--info c-alert--neutral margin-bottom-3"
    role="region"
    tabindex="-1"
  >
    <div
      class="usa-alert__body"
    >
      <h2
        class="usa-alert__heading font-heading-md text-bold"
        id="alert-heading1"
      >
        Only some people can apply online for now
      </h2>
      <div
        class="usa-alert__text"
      >
        <h3
          class="margin-top-3 font-heading-sm text-bold"
        >
          <svg
            class="margin-right-1 text-secondary text-middle margin-top-neg-05"
            fill="currentColor"
            height="20"
            viewBox="0 0 36 30"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M33 0H3v19.5h30zm-3 16.5H6V3h24zm3 4.5H3l-3 9h36zm-18.342 7.5l.699-1.5h5.286l.699 1.5z"
            />
          </svg>
          Apply online
        </h3>
        <p>
          If you are currently employed in Massachusetts but not self-employed, you can apply online or by phone for the following types of paid leave:
        </p>
        <ul
          class="usa-list"
        >
          <li>
            Medical leave due to your own illness, injury, or pregnancy
          </li>
          <li>
            Family leave to bond with your child after birth, adoption, or foster placement — whether you are applying before or after the child arrives
          </li>
          <li>
            Family leave to care for a family member with a serious medical condition
          </li>
        </ul>
        <h3
          class="font-heading-sm text-bold"
        >
          <svg
            class="margin-right-1 text-secondary text-middle margin-top-neg-05"
            fill="currentColor"
            height="20"
            viewBox="0 0 33 37"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20.841 35.51C9.798 40.57-7.343 7.19 3.451 1.56L6.609 0l5.235 10.23-3.121 1.54c-3.281 1.77 3.553 15.14 6.909 13.51.136-.06 3.083-1.51 3.094-1.52l5.278 10.2c-.012.01-2.982 1.47-3.163 1.55zm.399-29.23a9.49 9.49 0 016.376 11.81l-2.69-.8a6.681 6.681 0 00-.528-5.09 6.65 6.65 0 00-3.962-3.22zM19.838 11a4.542 4.542 0 012.71 2.21c.6 1.11.694 2.35.36 3.48l-4.373-1.31zm2.003-6.74c2.84.85 5.359 2.78 6.881 5.6 1.52 2.82 1.757 5.99.912 8.83l2.758.83a14.45 14.45 0 00-1.141-11.03 14.41 14.41 0 00-8.587-6.99z"
            />
          </svg>
          Apply by phone
        </h3>
        <p>
          Apply by calling the Department of Family and Medical Leave Contact Center at 
          <a
            href="tel:(833) 344-7365"
          >
            (833) 344‑7365
          </a>
           if 
          <strong>
            any
          </strong>
           of the following are true:
        </p>
        <ul
          class="usa-list"
        >
          <li>
            You are self-employed or unemployed and you are applying for paid leave
          </li>
          <li>
            You need paid family leave to care for a family member who serves in the armed forces
          </li>
          <li>
            You need paid family leave to manage family affairs when a family member is on active duty in the armed forces
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    class="measure-6"
  >
    <h2
      class="font-heading-md text-bold"
    >
      <svg
        aria-hidden="true"
        class="usa-icon margin-right-1 text-secondary text-middle margin-top-neg-05"
        focusable="false"
        role="img"
      >
        <use
          xlink:href="/img/sprite.svg#comment"
        />
      </svg>
      1. Tell your employer that you need to take paid leave
    </h2>
    <p>
      If you can, tell your employer at least 30 days before your leave begins. If you need to take leave right away, tell your employer as soon as possible.
    </p>
    <p>
      Once you tell your employer, you have the right to apply and your job is protected. Make a note of when you notified your employer. You will need to provide this date in your leave application.
    </p>
    <h2
      class="font-heading-md text-bold"
    >
      <svg
        aria-hidden="true"
        class="usa-icon margin-right-1 text-secondary text-middle margin-top-neg-05"
        focusable="false"
        role="img"
      >
        <use
          xlink:href="/img/sprite.svg#upload_file"
        />
      </svg>
      2. Get documentation that supports your leave request
    </h2>
    <p>
      If you’re not sure what kind of leave to apply for, 
      <a
        href="https://www.mass.gov/which-paid-leave"
        rel="noopener"
        target="_blank"
      >
        learn more about Paid Family and Medical Leave
      </a>
      .
    </p>
    <h3
      class="font-heading-sm text-bold"
    >
      Medical leave for your own serious health condition
    </h3>
    <p>
      Your health care provider must complete the 
      <a
        href="https://www.mass.gov/medical-leave-form"
        rel="noopener"
        target="_blank"
      >
        Certification of Your Serious Health Condition
      </a>
      .
    </p>
    <h3
      class="font-heading-sm text-bold"
    >
      Family leave
    </h3>
    <h4
      class="font-heading-xs text-bold"
    >
      To bond with a child
    </h4>
    <p>
      For family leave to bond with your child after birth, foster placement, or adoption you need a document that confirms your child’s date of birth or placement.
    </p>
    <p>
      You can apply before your child is born or arrives in your home. You will need to provide proof of birth or placement for your application to be approved.
    </p>
    <h4
      class="font-heading-xs text-bold"
    >
      To care for a family member with a serious health condition
    </h4>
    <p>
      The health care provider of the person you're caring for must complete the 
      <a
        href="https://www.mass.gov/family-caring-form"
        rel="noopener"
        target="_blank"
      >
        Certification of Your Family Member’s Serious Health Condition
      </a>
      .
    </p>
    <h2
      class="font-heading-md text-bold"
    >
      <svg
        aria-hidden="true"
        class="usa-icon margin-right-1 text-secondary text-middle margin-top-neg-05"
        focusable="false"
        role="img"
      >
        <use
          xlink:href="/img/sprite.svg#edit"
        />
      </svg>
      3. Apply
    </h2>
    <p>
      Applying takes around 15 minutes. Your information will save as you go, so you can finish your application later if you need to.
    </p>
    <p>
      If you give birth and plan to take both pregnancy-related medical leave and family leave to bond with your newborn, you should apply for medical leave first. Family leave to bond with your child can be 
      <a
        href="https://www.mass.gov/pfml-medical-bonding"
        rel="noopener"
        target="_blank"
      >
        easily added to your application
      </a>
       by calling our Contact Center at 
      <a
        href="tel:(833) 344-7365"
      >
        (833) 344‑7365
      </a>
      .
    </p>
    <p>
      You need to create multiple leave applications if you are:
    </p>
    <ul
      class="usa-list"
    >
      <li>
        Taking leave from multiple employers.
      </li>
      <li>
        Taking time off in uneven blocks of time (intermittent leave), 
        <strong>
          and
        </strong>
         taking time off completely or on a reduced schedule. You’ll need a separate application for the intermittent leave.
      </li>
    </ul>
    <p>
      PFML benefits are subject to reporting for tax purposes and may be subject to taxation. Your decision to have taxes withheld from your benefit may affect your personal tax liability. If you’re unsure whether you want to withhold taxes, we recommend speaking with a 
      <a
        href="https://mass.gov/taxassistance"
        rel="noopener"
        target="_blank"
      >
        tax professional
      </a>
      . Learn more about the 
      <a
        href="https://www.mass.gov/pfml-tax-liability"
        rel="noopener"
        target="_blank"
      >
        possible tax implications
      </a>
       of paid leave benefits.
    </p>
    <p>
      The maximum benefit a person can receive per week is $1084. Learn more about 
      <a
        href="https://www.mass.gov/how-much-will-i-get-paid"
        rel="noopener"
        target="_blank"
      >
        how benefits are calculated.
      </a>
    </p>
    <a
      class="usa-button margin-top-3 margin-bottom-3"
      href="/applications/start"
    >
      Create an application
    </a>
    <details
      class="c-details"
    >
      <summary
        class="margin-bottom-2 text-primary text-underline font-ui-xs"
      >
        Did you start an application by phone?
      </summary>
      <div
        class="margin-bottom-2"
      >
        <p>
          Some applications started on the phone can be added to this website. Once added, you can upload documents and see your application’s status and payments online.
        </p>
        <a
          class="display-inline-block margin-bottom-5"
          href="/applications/import-claim"
        >
          Add an existing application
        </a>
      </div>
    </details>
  </div>
</div>
`;
